File: 5 Advanced Algorithms and Complexity/Week 2 Linear Programming/Programming-Assignment-2/energy_values/energy_values.py
# python3
from decimal import Decimal, getcontext
from copy import deepcopy
getcontext().prec = 30
import math

EPS = 1e-6
PRECISION = 20

class Equation:
    def __init__(self, a, b):
        self.a = a
        self.b = b

class Position:
    def __init__(self, column, row):
        self.column = column
        self.row = row

def ReadEquation():
    size = int(input())
    a = []
    b = []
    for row in range(size):
        line = list(map(float, input().split()))
        a.append(line[:size])
        b.append(line[size])
    return Equation(a, b)

def SelectPivotElement(a, used_rows, used_columns):
    # This algorithm selects the first free element.
    # You'll need to improve it to pass the problem.
    pivot_element = Position(0, 0)
    while used_rows[pivot_element.row]:
        pivot_element.row += 1
    while used_columns[pivot_element.column]:
        pivot_element.column += 1
    
    maxValue = 0.0
    size = len(a)
    for i in range(pivot_element.row, size):
        if (math.fabs(a[i][pivot_element.column]) > math.fabs(maxValue)):
            maxValue = a[i][pivot_element.column]
            pivot_element.row = i
    
    return pivot_element

def SwapLines(a, b, used_rows, pivot_element):
    a[pivot_element.column], a[pivot_element.row] = a[pivot_element.row], a[pivot_element.column]
    b[pivot_element.column], b[pivot_element.row] = b[pivot_element.row], b[pivot_element.column]
    used_rows[pivot_element.column], used_rows[pivot_element.row] = used_rows[pivot_element.row], used_rows[pivot_element.column]
    pivot_element.row = pivot_element.column;

def BackSubstitution(a, b):
    size = len(a)
    for i in range(size-1, -1, -1):
        v = b[i]
        for j in range(0, i):
            b[j] -= a[j][i] * v
            a[j][i] = 0

def ScalePivot(a, b, pivot_element):
    divisor = a[pivot_element.row][pivot_element.column]
    size = len(a)
    
    for j in range(pivot_element.column, size):
        a[pivot_element.row][j] /= divisor
    
    b[pivot_element.row] /= divisor
    
def ProcessPivotElement(a, b, pivot_element):
    # Write your code here
    size = len(a)
    multiple = 0.0
    
    ScalePivot(a, b, pivot_element)
    
    for i in range(pivot_element.row + 1, size):
        multiple = a[i][pivot_element.column]
        for j in range(pivot_element.column, size):
            a[i][j] -= (a[pivot_element.row][j] * multiple)
        b[i] -= (b[pivot_element.row] * multiple)

def MarkPivotElementUsed(pivot_element, used_rows, used_columns):
    used_rows[pivot_element.row] = True
    used_columns[pivot_element.column] = True

def SolveEquation(equation):
    a = equation.a
    b = equation.b
    size = len(a)

    used_columns = [False] * size
    used_rows = [False] * size
    for step in range(size):
        pivot_element = SelectPivotElement(a, used_rows, used_columns)
        SwapLines(a, b, used_rows, pivot_element)
        ProcessPivotElement(a, b, pivot_element)
        MarkPivotElementUsed(pivot_element, used_rows, used_columns)
        
    BackSubstitution(a, b)

    return b

def PrintColumn(column):
    size = len(column)
    for row in range(size):
        print("%.20lf" % column[row])

if __name__ == "__main__":
    equation = ReadEquation()
    solution = SolveEquation(equation)
    PrintColumn(solution)
    exit(0)

File: 5 Advanced Algorithms and Complexity/Week 2 Linear Programming/Programming-Assignment-2/energy_values/energy_values.cpp
#include <cmath>
#include <iostream>
#include <vector>
#include <iomanip>

const double EPS = 1e-6;
const int PRECISION = 20;

typedef std::vector<double> Column;
typedef std::vector<double> Row;
typedef std::vector<Row> Matrix;

struct Equation {
    Equation(const Matrix &a, const Column &b):
        a(a),
        b(b)
    {}

    Matrix a;
    Column b;
};

struct Position {
    Position(int column, int row):
        column(column),
        row(row)
    {}

    int column;
    int row;
};

Equation ReadEquation() {
    int size;
    std::cin >> size;
    Matrix a(size, std::vector <double> (size, 0.0));
    Column b(size, 0.0);
    for (int row = 0; row < size; ++row) {
        for (int column = 0; column < size; ++column)
            std::cin >> a[row][column];
        std::cin >> b[row];
    }
    return Equation(a, b);
}

Position SelectPivotElement(
  const Matrix &a, 
  std::vector <bool> &used_rows, 
  std::vector <bool> &used_columns) {
    // This algorithm selects the first free element.
    // You'll need to improve it to pass the problem.
    Position pivot_element(0, 0);
    while (used_rows[pivot_element.row])
        ++pivot_element.row;
    while (used_columns[pivot_element.column])
        ++pivot_element.column;
    double max = 0.0;
    int size = a.size();
    for(int i = pivot_element.row; i < size; ++i){
        if (std::fabs(a[i][pivot_element.column]) > std::fabs(max)){
            max = a[i][pivot_element.column];
            pivot_element.row = i;
        }
    }
    return pivot_element;
}

void SwapLines(Matrix &a, Column &b, std::vector <bool> &used_rows, Position &pivot_element) {
    std::swap(a[pivot_element.column], a[pivot_element.row]);
    std::swap(b[pivot_element.column], b[pivot_element.row]);
    std::swap(used_rows[pivot_element.column], used_rows[pivot_element.row]);
    pivot_element.row = pivot_element.column;
}

void BackSubstitution(Matrix& a, Column& b){
    for (int i = a.size() - 1; i; --i){
        double v = b[i];
        for (int j = 0; j != i; ++j){
            b[j] -= a[j][i] * v;
            a[j][i] = 0;
        }
    }
}

inline void ScalePivot(Matrix& a, Column& b, const Position& pivot_element){
    const double divisor = a[pivot_element.row][pivot_element.column];
    const int size = a.size();
    
    for (int j = pivot_element.column; j < size; ++j){
        a[pivot_element.row][j] /= divisor;
    }
    
    b[pivot_element.row] /= divisor;
}

void ProcessPivotElement(Matrix &a, Column &b, const Position &pivot_element) {
    // Write your code here
    const int size = a.size();
    double multiple{0.0};
    
    ScalePivot(a, b, pivot_element);
    
    for (int i = pivot_element.row + 1; i < size; ++i){
        multiple = a[i][pivot_element.column];
        for (int j = pivot_element.column; j < size; ++j){
            a[i][j] -= (a[pivot_element.row][j] * multiple);
        }
        b[i] -= (b[pivot_element.row] * multiple);
    }
}

void MarkPivotElementUsed(const Position &pivot_element, std::vector <bool> &used_rows, std::vector <bool> &used_columns) {
    used_rows[pivot_element.row] = true;
    used_columns[pivot_element.column] = true;
}

Column SolveEquation(Equation equation) {
    Matrix &a = equation.a;
    Column &b = equation.b;
    int size = a.size();

    std::vector <bool> used_columns(size, false);
    std::vector <bool> used_rows(size, false);
    
    for (int step = 0; step < size; ++step) {
        Position pivot_element = SelectPivotElement(a, used_rows, used_columns);
        SwapLines(a, b, used_rows, pivot_element);
        ProcessPivotElement(a, b, pivot_element);
        MarkPivotElementUsed(pivot_element, used_rows, used_columns);
    }
    
    BackSubstitution(a, b);

    return b;
}

void PrintColumn(const Column &column) {
    int size = column.size();
    std::cout.precision(PRECISION);
    for (int row = 0; row < size; ++row)
        std::cout << column[row] << std::endl;
}

int main() {
    Equation equation = ReadEquation();
    Column solution = SolveEquation(equation);
    PrintColumn(solution);
    return 0;
}
